### PR TITLE
Add Redshift

### PR DESCRIPTION
**Pull request by manabusakai**
Hi, @dtan4.

Thank you for an awesome gem. I like this :smile:
This pull request to add the Redshift.

TODO:
- [x] Support other arguments
- [x] Add spec file

Please tell me if there is advice.

### CODE REVIEW

File: lib/terraforming/cli.rb [hunk 1]
```diff
@@ -120,6 +120,11 @@ def rds
       execute(Terraforming::Resource::RDS, options)
     end
 
+    desc "redshift", "Redshift"
+    def redshift
```
Comment (dtan4) [inline lib/terraforming/cli.rb]: `rs` (abbreviation of `Redshift`) is better :bulb:

Comment (manabusakai) [inline lib/terraforming/cli.rb]: Thanks :+1:
Fixed.

**Comment by manabusakai** [comment]
I removed [WIP] to PR title.

File: lib/terraforming/resource/redshift.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        @client = client
+      end
+
+      def tf
+        apply_template(@client, "tf/redshift")
+      end
+
+      def tfstate
+        clusters.inject({}) do |resources, cluster|
+          attributes = {
```
Comment (dtan4) [inline lib/terraforming/resource/redshift.rb]: `cluster_identifier` should be included in tfstate.
This causes re-create Redshift cluster by `terraform apply`:

```
-/+ aws_redshift_cluster.dtan4-hogefuga
    allow_version_upgrade:               "true" => "1"
    automated_snapshot_retention_period: "1" => "1"
    availability_zone:                   "ap-northeast-1a" => "ap-northeast-1a"
    cluster_identifier:                  "" => "dtan4-hogefuga" (forces new resource)
```

**Comment by dtan4** [comment]
We do not have specify `database_name` at the time of creating cluster.
If we do so, Redshift create default database named `dev`.

However, aws-sdk returns `database_name: nil` even if default database `dev` was created, and this causes the error below:

```bash
$ terraform plan
There are warnings and/or errors related to your configuration. Please
fix these before continuing.

Errors:

  * aws_redshift_cluster.dtan4-redshift: only lowercase letters characters allowed in "database_name"
  * aws_redshift_cluster.dtan4-redshift: "database_name" cannot be an empty string
```

```
# rs.tf
resource "aws_redshift_cluster" "dtan4-redshift" {
    cluster_identifier                  = "dtan4-redshift"
    database_name                       = ""
    # ...
}
```

To avoid this, Terraforming should exclude `database_name` from tf and tfstate if `database_name` is `nil`.

**Comment by manabusakai** [comment]
Thanks :bow:
Resolved the review.

**Comment by dtan4** [comment]
Please add one more thing (sorry) :exclamation: `skip_final_snapshot` attribute seems to be required.

Currently there is one diff with `terraform plan`:

```
$ terraform plan
Refreshing Terraform state prior to plan...

aws_redshift_cluster.dtan4-hoge: Refreshing state... (ID: dtan4-hoge)

The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

~ aws_redshift_cluster.dtan4-hoge
    skip_final_snapshot: "0" => "1"


Plan: 0 to add, 1 to change, 0 to destroy.
```

`skip_final_snapshot` can't be acquired via AWS SDK and is used with `terraform destroy`, so you only have to set the default value.

However, there is a difference between [document](https://www.terraform.io/docs/providers/aws/r/redshift_cluster.html) and [code](https://github.com/hashicorp/terraform/blob/1b3908633cbc6eb3cb7972550c0e875271b5feeb/builtin/providers/aws/resource_aws_redshift_cluster.go#L170) about the default value...
I created a issue https://github.com/hashicorp/terraform/issues/4925 about this. Please wait the official answer... :bow:

**Comment by dtan4** [comment]
According to https://github.com/hashicorp/terraform/issues/4925#issuecomment-177623314 , the default value of `skip_final_snapshot` is __true__.
@manabusakai Please use this value.

**Comment by manabusakai** [comment]
Sorry... :bow:
Please tell me the reason why `skip_final_snapshot` becomes `0`.

I tested it, but think that it becomes by default `1` and  always becomes `1`.

```
$ bundle exec terraforming rs | tee rs.tf
resource "aws_redshift_cluster" "develop" {
    cluster_identifier                  = "develop"
    database_name                       = "test"
    cluster_type                        = "multi-node"
    node_type                           = "dc1.large"
    master_password                     = "xxxxxxxx"
    master_username                     = "testuser"
    availability_zone                   = "ap-northeast-1c"
    preferred_maintenance_window        = "fri:15:00-fri:15:30"
    cluster_parameter_group_name        = "test"
    automated_snapshot_retention_period = "1"
    port                                = "5439"
    cluster_version                     = "1.0"
    allow_version_upgrade               = "true"
    number_of_nodes                     = "2"
    publicly_accessible                 = "true"
    encrypted                           = "true"
}
```

I set another AWS credentials.

```
$ terraform plan
Refreshing Terraform state prior to plan...

(snip)

+ aws_redshift_cluster.develop
    allow_version_upgrade:               "" => "1"
    automated_snapshot_retention_period: "" => "1"
    availability_zone:                   "" => "ap-northeast-1c"
    cluster_identifier:                  "" => "develop"
    cluster_parameter_group_name:        "" => "test"
    cluster_public_key:                  "" => "<computed>"
    cluster_revision_number:             "" => "<computed>"
    cluster_security_groups.#:           "" => "<computed>"
    cluster_subnet_group_name:           "" => "<computed>"
    cluster_type:                        "" => "multi-node"
    cluster_version:                     "" => "1.0"
    database_name:                       "" => "test"
    encrypted:                           "" => "1"
    endpoint:                            "" => "<computed>"
    master_password:                     "" => "xxxxxxxx"
    master_username:                     "" => "testuser"
    node_type:                           "" => "dc1.large"
    number_of_nodes:                     "" => "2"
    port:                                "" => "5439"
    preferred_maintenance_window:        "" => "fri:15:00-fri:15:30"
    publicly_accessible:                 "" => "1"
    skip_final_snapshot:                 "" => "1"
    vpc_security_group_ids.#:            "" => "<computed>"


Plan: 1 to add, 0 to change, 0 to destroy.
```

Is `skip_final_snapshot` really required?

**Comment by manabusakai** [comment]
ping @dtan4

Let me know if you need any help.

**Comment by dtan4** [comment]
I'm deeply sorry for late reply...

With Terraform v0.6.11, `skip_final_snapshot` becomes empty. Here is the test result:

```bash
$ be bin/terraforming rs | tee tmp/rs.tf
resource "aws_redshift_cluster" "dtan4-hoge" {
    cluster_identifier                  = "dtan4-hoge"
    cluster_type                        = "single-node"
    node_type                           = "dc1.large"
    master_password                     = "xxxxxxxx"
    master_username                     = "hoge"
    availability_zone                   = "ap-northeast-1c"
    preferred_maintenance_window        = "tue:15:30-tue:16:00"
    cluster_parameter_group_name        = "default.redshift-1.0"
    automated_snapshot_retention_period = "1"
    port                                = "5439"
    cluster_version                     = "1.0"
    allow_version_upgrade               = "true"
    number_of_nodes                     = "1"
    publicly_accessible                 = "true"
    encrypted                           = "false"
}

$ be bin/terraforming rs --tfstate | tee tmp/terraform.tfstate
{
  "version": 1,
  "serial": 1,
  "modules": [
    {
      "path": [
        "root"
      ],
      "outputs": {
      },
      "resources": {
        "aws_redshift_cluster.dtan4-hoge": {
          "type": "aws_redshift_cluster",
          "primary": {
            "id": "dtan4-hoge",
            "attributes": {
              "cluster_identifier": "dtan4-hoge",
              "cluster_type": "single-node",
              "node_type": "dc1.large",
              "master_password": "xxxxxxxx",
              "master_username": "hoge",
              "availability_zone": "ap-northeast-1c",
              "preferred_maintenance_window": "tue:15:30-tue:16:00",
              "cluster_parameter_group_name": "default.redshift-1.0",
              "automated_snapshot_retention_period": "1",
              "port": "5439",
              "cluster_version": "1.0",
              "allow_version_upgrade": "true",
              "number_of_nodes": "1",
              "publicly_accessible": "true",
              "encrypted": "false"
            }
          }
        }
      }
    }
  ]
}
```

```
$ ~/Downloads/terraform version
Terraform v0.6.11

$ ~/Downloads/terraform plan
Refreshing Terraform state prior to plan...

aws_redshift_cluster.dtan4-hoge: Refreshing state... (ID: dtan4-hoge)

The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

~ aws_redshift_cluster.dtan4-hoge
    skip_final_snapshot: "" => "1"


Plan: 0 to add, 1 to change, 0 to destroy.
```

However, `skip_final_snapshot` is still required not to make diff.

Modification of `skip_final_snapshot` does not affect to real Redshift cluster, but Terraforming should generate tf and tfstate without any diff with `terraform plan`, I think.

**Comment by manabusakai** [comment]
Thank you for the reply!
I understood that you think :+1:

Push the code.

**Comment by dtan4** [comment]
It works fine :exclamation: 

```bash
$ ~/Downloads/terraform plan
Refreshing Terraform state prior to plan...

aws_redshift_cluster.dtan4-hoge: Refreshing state... (ID: dtan4-hoge)

No changes. Infrastructure is up-to-date. This means that Terraform
could not detect any differences between your configuration and
the real physical resources that exist. As a result, Terraform
doesn't need to do anything.
```

LGTM :+1: :+1: Thanks :tada: :exclamation:

**Comment by manabusakai** [comment]
Thanks! :smile:

**Pull request merged**